packages/superset-ui-control-utils/src/types.ts

```diff
@@ -198,21 +193,44 @@ type SelectControlType =
   | 'SelectControlVerifiedOptions'
   | 'AdhocFilterControlVerifiedOptions';
 
+// via react-select/src/filters
+interface FilterOption<T extends SelectOption> {
+  label: string;
+  value: string;
+  data: T;
+}
+
+// Ref: superset-frontend/src/components/Select/SupersetStyledSelect.tsx
 export interface SelectControlConfig<T extends SelectOption = SelectOption>
-  extends GeneralControlConfig {
+  extends BaseControlConfig {
   type: SelectControlType;
-  options?: T[];
   clearable?: boolean;
   freeForm?: boolean;
   multi?: boolean;
-  optionRenderer?: (option: T) => ReactNode;
-  valueRenderer?: (option: T) => ReactNode;
   valueKey?: string;
   labelKey?: string;
+  options?: T[];
+  optionRenderer?: (option: T) => ReactNode;
+  valueRenderer?: (option: T) => ReactNode;
+  filterOption?: ((option: FilterOption<T>, rawInput: string) => Boolean) | null;
 }
 
+export type SharedControlConfig<
+  T extends InternalControlType = InternalControlType,
+  O extends SelectOption = SelectOption
+> = T extends SelectControlType ? SelectControlConfig<O> : BaseControlConfig & { type: T };
+
+/** --------------------------------------------
+ * Custom controls
+ * --------------------------------------------- */
+export type CustomComponentControlConfig<P = unknown> = BaseControlConfig & {
+  type: InternalControlType | React.ComponentType<P>;
+} & Omit<P, 'onChange' | 'hovered'>; // two run-time properties from superset-frontend/src/explore/components/Control.jsx
+
+// Catch-all ControlConfig
 export type ControlConfig<T extends SelectOption = SelectOption> =
-  | GeneralControlConfig
+  | BaseControlConfig
```

Comment:
Do I understand correctly that a control config is either `SharedControlConfig` or `CustomComponentControlConfig`? 
Should `BaseControlConfig` then be removed from this catch-all? 

```ts
export type ControlConfig<
  T extends InternalControlType = InternalControlType
  O extends SelectOption = SelectOption
> = 
  | SharedControlConfig<T, O>
  | CustomComponentConfig;
```

Reply:
True, this needs an update.

BTW, got this error when removed `BaseControlConfig` from the export:

```
Default export of the module has or is using private name 'BaseControlConfig'.
```

It seems typescript requires you to export all internal types used by a default export.